docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+parsed at compile time and generate the correct stubs for the interoperability
+layer to correctly implement the remote procedure call (RPC).
+
+Each execution model (child process, web assembly, etc) will need explicit
+support for deploying the functionality at run time, hopefully not individually
+bound to the source language (N-to-M).
+
+But the overall architecture will be the same.
+
+### Parser
+
+1. Upon finding a file include that has foreign origin, the appropriate foreign
+   language parser (clang for C++) will be asked to parse the file and generate
+   its own AST.
+2. A Verona module will be created, with all foreign builtin types.
+3. When the Verona compiler asks for a particular type or function in the foreign
+   module, the language parser will convert the signature of its own AST into
+   Verona AST and add that node in its own module, then pass that to the Verona
+   AST node.
+
+This allows lazy evaluation of what could be very large number of includes, with
+code that can't be translated into Verona.
+
+The end result is an AST that has foreign nodes as Verona nodes on Verona modules
+without the actual implementation. For all purposes, this is a plain Verona AST
+with calls to functions that have been declared, but not defined.
+
+Definition (and implementation in object code) will be done later by the compiler
+once we know all functions that have survived the initial round of cleanups (ex.
+reachability analysis).
+
+### Compiler
+
+1. For each sandbox region, the compiler generates calls to the runtime to setup
+   the region, including its allocators.
+2. Upon finding a foreign call, the compiler will ask the language driver for
+   the foreign AST node, to implement the actual call.
+3. The driver will generate the conversion between Verona and native types,
+   including structure layout handling per architecture, serialise the
+   arguments and return values into a buffer and call the dispatcher below.
+4. A dispatcher will be created to call those functions with serialised
+   arguments by an index that is associated when they're created.
+5. A bi-directional RPC layer is created between the Verona code and the sandbox,
+   where calls to external functions go one way, and calls to allocation, system
+   calls, etc. go the other way.
+6. The appropriate object files are created and linked with the main executable.
+
+Not all functions will be representable in Verona, not all structure layouts will
+be easily known at the Verona level (special hardware ABI), so errors may still
+occur here.
+
+The compiler is responsible for emitting errors that inform the programmer what
+type/function cannot be represented and potentially why. The programmer is
+responsible for implementing a wrapper function that abstracts away the
+complexity of the language in a public function that Verona can call.
+
+Language support will be always evolving. This is expected to be a balance
+between "best effort" and "appropriate support".
```

Comment:
In particular, this should be test driven: We should document what works and we should only implement things needed to get examples to work.  Once we can do MLIR things in Verona, we can probably declare success.

Reply:
I think there will always be things we don't implement. I think success will be different depending on what you mean. v1 success is making it work at all with something non-trivial. Language success is not having to rely on ugly wrappers (like Java's JNI) for every little trip outside of the Verona's confines.